### ISSUE TITLE
Research how should CFDP work

### ISSUE THREAD

**Issue by Qlas** [uncommon-nickname/cfdp#20]
CCSDS documents:
- [[BLUE BOOK] CCSDS FILE DELIVERY PROTOCOL (CFDP)](https://public.ccsds.org/Pubs/727x0b5.pdf)
- [[GREEN BOOK] PART 1: INTRODUCTION AND OVERVIEW](https://public.ccsds.org/Pubs/720x1g4.pdf)
- [[GREEN BOOK] PART 2: IMPLEMENTERS GUIDE](https://public.ccsds.org/Pubs/720x2g4.pdf)
- [[YELLOW BOOK] NOTEBOOK OF COMMON INTER-AGENCY TESTS FOR CORE PROCEDURES](https://public.ccsds.org/Pubs/720x4y1.pdf)

We should support:
- Service classes
  - CLASS 1 (UNRELIABLE TRANSFER)
    - Without Transaction closure requested (should be possible via simplex transmission)
    - With Transaction closure requested (requires duplex transmission)
  - CLASS 2 (RELIABLE TRANSFER) (requires duplex transmission)
    - NACK modes (more info: [[GREEN BOOK] PART 2: IMPLEMENTERS GUIDE](https://public.ccsds.org/Pubs/720x2g4.pdf) 2.6.3.1)
      - Deferred
      - Immediate
      - Prompted
      - Asnychronous  
  - ~~CLASS 3 (UNRELIABLE TRANSFER VIA MULTIPLE WAYPOINTS)~~
  - ~~CLASS 4 (RELIABLE TRANSFER VIA MULTIPLE WAYPOINTS)~~
> New CFDP specification omits classes 3 and 4

- User operations 
  - Proxy
    - Proxy Put Request
    - Proxy Message to User
    - Proxy Filestore Request
    - Proxy Fault Handler Override
    - Proxy Transmission Mode
    - Proxy Flow Label
    - Proxy Segmentation Control
    - Proxy Put Response
    - Proxy Filestore Response
    - Proxy Put Cancel
    - Proxy Closure Request
  - Remote status report
    - Remote Status Report Request
    - Remote Status Report Response
  - Directory 
    - Directory Listing Request
    - Directory Listing Response
  - Remote suspend
    - Remote Suspend Request
    - Remote Suspend Response
  - Remote resume
    - Remote Resume Request
    - Remote Resume Response
> User Operations refer to the use of the CFDP services offered by the local CFDP
entity to cause the CFDP user of a remote CFDP entity to initiate additional CFDP
transactions.

- Communication between CFDP User and its local enitity via primitives:
![image](https://github.com/user-attachments/assets/6938c430-4abb-4847-b7b4-c8d33e6e648e)

- Checksum algorithms:
  - Null (no checksum)
  - Modular checksum
  - In addition, we can implement others from the [SANA Checksum Identifiers registry](https://sanaregistry.org/r/checksum_identifiers/)

- Transport
  - Layer(s) 
    - UDP
    - Any type of TCP
  - Should we provide Space Packet implementation on top of transport layers?
- CFDP User
![image](https://github.com/user-attachments/assets/f094390d-aa69-45e0-83ea-e55ac9ebcbf8)

**Comment by Qlas** [comment]
There is also a set of timers and counters that should be implemented:
![image](https://github.com/user-attachments/assets/e95b93b0-bc0b-45c5-826f-776028aa5f7f)
![image](https://github.com/user-attachments/assets/fa3351b3-5ddc-46ec-a4b0-5ba63d36df56)

**Comment by Qlas** [comment]
The main part of communication between entities is **Transaction**. Transaction should be saved on both (receiver and sender) Entities with individual timers and counters. T has unique ID (4 bytes).
> Should we preserve information about ongoing transaction between runs of the CFDP daemon?

One transaction can only transmit only one FDU (which can be empty -> only metadata FDU)

During a transaction, there are usually multiple PDUs that contain a piece of FDU or metadata. 
There is 2 types of PDU:
- File Directive
  - EOF PDU
  -  Finished PDU
  - ACK PDU
  - Metadata PDU
    - Can contains `MessageToUser` and `Filestore`  requests
  - NAK PDU
  - Prompt PDU
  - Keep Alive PDU
- File Data

`MessageToUser` and `Filestore` requests have to been transmitted on the beginning of the transaction
> **IMPORTANT**: `MessageToUser` requests should be send immediately to the Remote CFDP User via `Metadata-Recv.indication` but `Filestore`  requests have to be sent only after all data transfer activities (if any)

Only one Metadata PDU can be send during transaction

**Comment by Qlas** [comment]
![new-pdu-processing](https://github.com/user-attachments/assets/d7f61a11-1e9a-414e-9d9a-f62cb88f3668)
